### DISCUSSION TITLE
Idiomatic Python API

### DISCUSSION THREAD

**Discussion by rgov** [eProsima/Fast-DDS-python#62] [Ideas]
Has any thought been given to developing an idiomatic Python API for Fast DDS? The API requires a coding style that is fairly unnatural in Python.

Some examples:

Accessing instance variables on objects:

```python
# not idiomatic
data.message()
self.topic_data_type.getName()

# idiomatic
data.message
self.topic_data_type.name
```

Setting instance variables on objects:

```python
# not idiomatic
data.message("Hello World")
self.topic_data_type.setName("HelloWorld")  # p.s., why does this one use "set" and other doesn't?


# idiomatic
data.message = "Hello World"
self.topic_data_type.name = "HelloWorld"
```

Passing objects by reference to have them initialized: 

```python
# not idiomatic
info = fastdds.SampleInfo()
data = HelloWorld.HelloWorld()
reader.take_next_sample(data, info)

# idiomatic
data, info = reader.take_next_sample()


# not idiomatic
self.subscriber_qos = fastdds.SubscriberQos()
self.participant.get_default_subscriber_qos(self.subscriber_qos)

# idiomatic, kind of?
self.subscriber_qos = self.participant.create_subscriber_qos()
self.subscriber_qos = fastdds.SubscriberQos.create_from(self.participant)
```

Manual management of resources:

```python
# not idiomatic
factory = fastdds.DomainParticipantFactory.get_instance()
self.participant.delete_contained_entities()
factory.delete_participant(self.participant)

# idiomatic, often happens implicitly
del self.participant
```

**Comment by rgov** [comment]
Presumably this is mostly a consequence of using SWIG to generate the Python bindings.

SWIG does have [`%attribute`](https://www.swig.org/Doc4.1/Library.html#Library_attributes) for wrapping setters and getters in an idiomatic way.

I've also heard decent things about [pybind11](https://pybind11.readthedocs.io/en/stable/index.html). [This example](https://pybind11.readthedocs.io/en/stable/classes.html#instance-and-static-fields) shows how to handle mapping Python object properties to C++ setters and getters like `getName()` and `setName()`

**Comment by JLBuenoLopez** [comment]
Thanks @rgov for your contribution

I have labeled this as an enhancement and I am moving the ticket to the Ideas discussion forum. For the moment the Python API will remain as it is.
